config/v1/types_ingress.go

```diff
@@ -41,13 +44,103 @@ type IngressSpec struct {
+	// can customize hostnames and serving certificates.
+	//
+	// An operator that creates entries in this slice should clean them up during removal.
+	// An operator must also handle the case of deleted status without churn.
+	// +kubebuilder:validation:Optional
+	// +optional
+	ComponentRoutes []ComponentRouteStatus `json:"componentRoutes,omitempty"`
 }
 
-// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// ComponentRouteSpec allows for configuration of a route's hostname and serving certificate.
+type ComponentRouteSpec struct {
+	// Namespace is the namespace of the route to customize. It must be a real namespace. Using an actual namespace
+	// ensures that no two components will conflict and the same component can be installed multiple times.
+	// +kubebuilder:validation:Pattern=[a-z0-9]([-a-z0-9]*[a-z0-9])?
+	// +kubebuilder:validation:Required
+	// +required
+	Namespace string `json:"namespace"`
+
+	// Name is the logical name of the route to customize.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+
+	// Hostname is the hostname that should be supported by the route.
+	// +kubebuilder:validation:Required
+	// +required
+	Hostname string `json:"hostname"`
+
+	// ServingCertKeyPairSecret is a reference to a secret in namespace/openshift-config that is a kubernetes tls secret.
+	// The serving cert/key pair must match and will be used by the operator to fulfill the intent of serving with this name.
+	// +kubebuilder:validation:Required
+	// +required
+	ServingCertKeyPairSecret SecretNameReference `json:"servingCertKeyPairSecret"`
+}
+
+// ComponentRouteStatus contains information allowing configuration of a route's hostname and serving certificate.
+type ComponentRouteStatus struct {
+	// Namespace is the namespace of the route to customize. It must be a real namespace. Using an actual namespace
+	// ensures that no two components will conflict and the same component can be installed multiple times.
+	// +kubebuilder:validation:Pattern=[a-z0-9]([-a-z0-9]*[a-z0-9])?
+	// +kubebuilder:validation:Required
+	// +required
+	Namespace string `json:"namespace"`
 
+	// Name is the logical name of the route to customize. It does not have to be the actual name of a route resource.
+	// Keep in mind that this is your API for users to customize. You could later rename the route, but you cannot rename
+	// this name.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+
+	// DefaultHostname is the normal hostname of this route.
+	// +kubebuilder:validation:Required
+	// +required
+	DefaultHostname string `json:"defaultHostname"`
+
+	// ConsumingUsers is a slice of ServiceAccounts that need to have read permission on the ServingCertKeyPairSecret secret.
+	// +kubebuilder:validation:Required
+	// +required
+	ConsumingUsers []string `json:"consumingUsers"`
```

Comment:
can this be empty?

Reply:
The only reason it could be empty is if the operator already has read permissions in the `openshift-config` namespace. It seems fine to force a minimum length of 1 to prevent confusion rather than support this case.

Reply:
not sure. This is the status, i.e. written by operators. IMO we can make this optional and unconstrained. it only drives the controller, right?

Reply:
You are correct, [the plan](https://github.com/openshift/enhancements/pull/577/files#r573785519) is to have this field set by an operator that opts into this feature and is used by the RBAC controller to identify the service account that should be included in the rolebinding.

Reply:
I have removed all validation and marked it as optional as requested.

Reply:
@sttts come to think of it, given [this conversation](https://github.com/openshift/enhancements/pull/577/files#r573187911), should we add regex for valid service account names?

Reply:
@sttts come to think of it, given [this conversation](https://github.com/openshift/enhancements/pull/577/files#r573187911), should we add regex for valid service account names, i.e., `system:serviceaccount:<namespace>:<sa-name>`?

Reply:
> I have removed all validation and marked it as optional as requested.

you need length validation.  Let's say that 512 characters is enough for anyone

Reply:
and lets say that no more than 5 users.

Reply:
I will limit the number of characters to 512 and set a max array size of 5. Per [our earlier conversation](https://github.com/openshift/enhancements/pull/577/files#r573187911), there is a requirement that the operator provide the full `system:serviceaccount:<namespace>:<sa-name>`, should we add any regex validation or just expect the operators to provide the correct name?

Reply:
if that's our opening restriction, yes.  That sounds like past me.

Reply:
still waiting on this

Reply:
Done!